lib/spec_rewrite.ts

```diff
@@ -169,26 +176,105 @@ export function dataRewrite(tableName: string, transform: Transforms, db: string
       as: "bin1"
     })
 
-    return true
+    return true // skip the next aggregate transform
   }
 
   if (transform.type === "aggregate") {
+    var prev = dbTransforms.pop() ?? null // null or a dbtransform
+
+    dbTransforms.push({
+      type: "dbtransform",
+      name: transform.name,
+      query: {
+        signal: aggregateTransformToSql(tableName, transform, db, prev)
+      }
+    })
+  }
+
+  if (transform.type === "filter") {
+    var prev = dbTransforms.pop() ?? null // null or a dbtransform
+
     dbTransforms.push({
       type: "dbtransform",
+      name: transform.name,
       query: {
-        signal: aggregateTransformToSql(tableName, transform, db)
+        signal: filterTransformToSql(tableName, transform, db, prev)
       }
     })
   }
 
 }
+const filterTransformToSql = (tableName: string, transform: any, db: string, prev: any) => {
+  console.log(parse(transform.expr))
+  const filter = expr2sql(parse(transform.expr))
+  tableName = prev ? `(${prev.query.signal.slice(1, -1)}) ${prev.name}` : tableName
+
+
+  var sql = ''
+  sql = [
+    `SELECT *`,
+    `FROM ${tableName}`,
+    `WHERE ${filter}`
+  ].join(" ")
+
+  return `"${sql}"`
+}
+
+function expr2sql(expr) {
+  var memberDepth = 0
+  function visit(ast) {
+    const generator = Generators[ast.type];
+    return generator(ast);
+  }
+  const Generators = {
+    Literal: n => n.raw,
+
+    Identifier: n => {
+      const id = n.name;
+      if (memberDepth > 0) {
+        return id;
+      }
+    },
+
+    MemberExpression: n => {
+      const d = !n.computed,
+        o = visit(n.object);
+      if (d) memberDepth += 1;
+      const p = visit(n.property);
+
+      if (d) memberDepth -= 1;
+      return p;
+    },
+
+    BinaryExpression: n => {
+      const right = visit(n.right)
+      if (right === 'null') {
+        n.operator = (n.operator === '==' || n.operator === '===') ? 'IS' : 'IS NOT'
+      } else {
+        if (n.operoter === '==' || n.operoter === '===') {
+          n.operator = '='
+        } else if (n.operator === '!=' || n.operator === '!==') {
+          n.operator = '!='
+        }
+      }
+      return visit(n.left) + ' ' + n.operator + ' ' + right
+    },
+
+    LogicalExpression: n => {
+      n.operator = n.operator === '&&' ? 'AND' : 'OR'
+      return visit(n.left) + ' ' + n.operator + ' ' + visit(n.right)
+    },
+  }
+  return visit(expr)
+}
 
 export function specRewrite(vgSpec) {
```

Comment:
I'm a bit confused here. Did we decide to rewrite the spec or the dataflow? This seems to be doing the former.

Reply:
We decided to rewrite the spec. Andrew's previous version was rewriting the dataflow.